File: web/src/lib/repositories/supabase-vehicle-repository.ts
import { SupabaseClient } from '@supabase/supabase-js';
import type { Vehicle, CreateVehicleDTO, VehicleRepository } from './vehicle-repository';

export class SupabaseVehicleRepository implements VehicleRepository {
  constructor(private supabase: SupabaseClient) {}

  async getAll(): Promise<Vehicle[]> {
    const { data, error } = await this.supabase
      .from('vehicles')
      .select('*')
      .order('created_at', { ascending: false });

    if (error) throw new Error(error.message);
    return data as Vehicle[];
  }

  async getById(id: string): Promise<Vehicle | null> {
    const { data, error } = await this.supabase
      .from('vehicles')
      .select('*')
      .eq('id', id)
      .single();

    if (error) {
      if (error.code === 'PGRST116') return null;
      throw new Error(error.message);
    }
    return data as Vehicle;
  }

  async create(dto: CreateVehicleDTO): Promise<Vehicle> {
    const { data, error } = await this.supabase
      .from('vehicles')
      .insert(dto)
      .select()
      .single();

    if (error) throw new Error(error.message);
    return data as Vehicle;
  }

  async update(id: string, dto: Partial<CreateVehicleDTO>): Promise<Vehicle> {
    const { data, error } = await this.supabase
      .from('vehicles')
      .update(dto)
      .eq('id', id)
      .select()
      .single();

    if (error) throw new Error(error.message);
    return data as Vehicle;
  }

  async delete(id: string): Promise<void> {
    const { error } = await this.supabase
      .from('vehicles')
      .delete()
      .eq('id', id);

    if (error) throw new Error(error.message);
  }
}

File: mobile/lib/features/vehicles/data/supabase_vehicle_repository.dart
import 'package:supabase_flutter/supabase_flutter.dart';
import '../domain/vehicle.dart';

/// Defines the standard contract for Vehicle data operations.
/// 
/// This repository is implemented by [SupabaseVehicleRepository] for production
/// and can be mocked for testing. It interacts directly with the `Vehicle`
/// domain model.
abstract class VehicleRepository {
  /// Fetches all vehicles registered to the current tenant, ordered by creation date descending.
  Future<List<Vehicle>> getAll();

  /// Fetches a specific vehicle by its UUID. Returns null if not found.
  Future<Vehicle?> getById(String id);

  /// Creates a new vehicle record. Returns the newly created [Vehicle] with its generated UUID.
  Future<Vehicle> create(Vehicle vehicle);

  /// Updates specific fields of an existing vehicle.
  Future<void> update(String id, Map<String, dynamic> data);

  /// Hard deletes a vehicle from the system by its UUID.
  Future<void> delete(String id);
}

/// Production implementation of [VehicleRepository] using Supabase.
/// 
/// Relies on the `vehicles` table. Assumes that Row Level Security (RLS) policies 
/// automatically scope data requests to the authenticated user's current tenant.
class SupabaseVehicleRepository implements VehicleRepository {
  final SupabaseClient _supabase;

  SupabaseVehicleRepository(this._supabase);

  @override
  Future<List<Vehicle>> getAll() async {
    final response = await _supabase
        .from('vehicles')
        .select('*')
        .order('created_at', ascending: false);
    
    return (response as List).map((json) => Vehicle.fromJson(json)).toList();
  }

  @override
  Future<Vehicle?> getById(String id) async {
    final response = await _supabase
        .from('vehicles')
        .select('*')
        .eq('id', id)
        .maybeSingle();

    if (response == null) return null;
    return Vehicle.fromJson(response);
  }

  @override
  Future<Vehicle> create(Vehicle vehicle) async {
    final response = await _supabase
        .from('vehicles')
        .insert(vehicle.toJson())
        .select()
        .single();

    return Vehicle.fromJson(response);
  }

  @override
  Future<void> update(String id, Map<String, dynamic> data) async {
    await _supabase
        .from('vehicles')
        .update(data)
        .eq('id', id);
  }

  @override
  Future<void> delete(String id) async {
    await _supabase
        .from('vehicles')
        .delete()
        .eq('id', id);
  }
}
